File: Reqnroll/Chapter17/17.2-InPlacePersonaPattern/Before/WIMP.Specs/StepDefinitions/OrderingStepDefinitions.cs
using Reqnroll;

using WIMP.App.Models;
using WIMP.Specs.Drivers;
using WIMP.Specs.Support;

namespace WIMP.Specs.StepDefinitions;

[Binding]
public class OrderingStepDefinitions(
    OrderingApiDriver orderingApiDriver,
    BackdoorApiDriver backdoorApiDriver)
{
    [Given("the customer has the following orders")]
    public async Task GivenTheCustomerHasTheFollowingOrders(DataTable ordersTable)
    {
        var orders = ordersTable.CreateSet<OrderByNumberData>().ToList();

        foreach (var order in orders)
        {
            var orderRequest = new PlaceOrderRequestObjectMother().Build();
            await backdoorApiDriver
                .PrepareOrder(DomainDefaults.CustomerName, orderRequest, order.Status)
                .Execute();
        }
    }

    [When("they cancel {order}")]
    public async Task WhenTheyCancelOrder(int orderNo)
    {
        await orderingApiDriver.CancelOrder(orderNo).Execute();
    }

    [Then("their order list should contain")]
    public async Task ThenTheirOrderListShouldContain(DataTable expectedOrdersTable)
    {
        var expectedOrders = expectedOrdersTable.CreateSet<OrderByNumberData>().ToList();
        foreach (var expectedOrder in expectedOrders)
        {
            var order = await orderingApiDriver.GetOrder(expectedOrder.OrderNo).Execute();
            Assert.AreEqual(expectedOrder.Status, order.Status,
                $"Unexpected status for order {expectedOrder.OrderNo}.");
        }
    }
}

public class OrderByNumberData
{
    public int OrderNo { get; set; }
    public OrderStatus Status { get; set; }
}

File: CucumberJVM/Chapter17/17.2-InPlacePersonaPattern/Before/src/test/java/com/wimp/app/specs/stepdefinitions/OrderingStepDefinitions.java
package com.wimp.app.specs.stepdefinitions;

import com.wimp.app.specs.drivers.BackdoorApiDriver;
import com.wimp.app.specs.drivers.OrderingApiDriver;
import com.wimp.app.specs.support.DomainDefaults;
import com.wimp.app.specs.support.OrderByNumberData;
import com.wimp.app.specs.support.PlaceOrderRequestObjectMother;
import io.cucumber.java.en.*;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class OrderingStepDefinitions {
    private final OrderingApiDriver orderingApiDriver;
    private final BackdoorApiDriver backdoorApiDriver;

    public OrderingStepDefinitions(OrderingApiDriver orderingApiDriver, BackdoorApiDriver backdoorApiDriver) {
        this.orderingApiDriver = orderingApiDriver;
        this.backdoorApiDriver = backdoorApiDriver;
    }

    @Given("the customer has the following orders")
    public void theCustomerHasTheFollowingOrders(List<OrderByNumberData> orders) throws Exception {
        for (final OrderByNumberData order : orders) {
            var orderRequest = new PlaceOrderRequestObjectMother().build();
            backdoorApiDriver
                .prepareOrder(DomainDefaults.CUSTOMER_NAME, orderRequest, order.getStatus())
                .execute();
        }
    }

    @When("they cancel {order}")
    public void theyCancelOrder(int orderNo) throws Exception {
        orderingApiDriver.cancelOrder(orderNo).execute();
    }

    @Then("their order list should contain")
    public void thenTheirOrderListShouldContain(List<OrderByNumberData> expectedOrders) throws Exception {
        for (final OrderByNumberData expectedOrder : expectedOrders) {
            var order = orderingApiDriver.getOrder(expectedOrder.getOrderNo()).execute();
            assertEquals(expectedOrder.getStatus(), order.getStatus(), "Unexpected status for order %s.".formatted(expectedOrder.getOrderNo()));
        }
    }
}
